### PR TITLE
#4438 Encapsulate Axios as observers that rxjs can use

### PR DESCRIPTION
**Pull request by wu87988622**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4438 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/store/epics/axiosObservable.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+export const axiosObservable = api =>
+  Observable.create(observer => {
+    api
+      .then(o => {
+        observer.next(o);
+        observer.complete();
+      })
+      .catch(e => observer.error(e));
+  });
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/axiosObservable.js]: 👍

Comment (eechih) [reply ohara-manager/client/src/store/epics/axiosObservable.js]: 我剛剛實驗發現，可以直接用 defer 將 axios api 裝起來，效果等同於 Observable.create。

```
const waitWorker$ = params =>
  defer(() => workerApi.get(params)).pipe(
    map(res => res.data),
    map(data => {
      if (!data?.state || data.state !== SERVICE_STATE.RUNNING) {
        throw data;
      }
      return actions.startWorker.success(data);
    }),
    retryWhen(error => error.pipe(delay(duration * 2), take(retry))),
    catchError(error => of(actions.startWorker.failure(error))),
  );	  
);
```

Comment (wu87988622) [reply ohara-manager/client/src/store/epics/axiosObservable.js]: 我剛剛實驗一下確實，那就以後只需要將from=>defer就可以了，不需要自己定義了。

File: ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js [hunk 2]
```diff
@@ -38,148 +38,140 @@ import * as actions from 'store/actions';
...
+        createWorker$(action),
+        startZookeeper$(action),
+        checkZookeeper$(action),
+        startBroker$(action),
+        checkBroker$(action),
+        startWorker$(action),
+        checkWorker$(action),
+        createWorkspaceSuccess(action),
+        catchError(res => of(action.createWorkspace.failure(res))),
+      ).pipe(concatAll()),
```
Comment (eechih) [inline ohara-manager/client/src/store/epics/workspace/createWorkspaceEpic.js]: @wu87988622 我已參考你的做法重新寫過 createWorkapceEpic (在 PR #4450)，你可以幫忙看一下嗎？ 3Q

**Comment by chia7712** [comment]
retry

**Comment by chia7712** [comment]
retry

**Comment by eechih** [comment]
@wu87988622 你好像忘了移除 node context & api？或是在其他 issue 處理？

**Comment by wu87988622** [comment]
> @wu87988622 你好像忘了移除 node context & api？或是在其他 issue 處理？

ok i forget this ~ thanks

**Pull request merged**